Smp pairing both sides handling

joerchan opened the pull request:
Handle both Central and Peripheral initiating security at the same time, this would result in the host starting the encryption procedure twice in the controller, one of them will then fail.

joerchan committed:
Bluetooth: SMP: Handle both devices initiating security

Handle case where:
 - Peripheral sends security request after master has sent pairing
   request or started encryption procedure.
   This packet can be ignored, as long as the slave has not already
   responded with pairing response.
 - Central wants to start security after peripheral initiated security
   request, return error code busy in this case

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3211,13 +3211,33 @@ static void le_ltk_neg_reply(u16_t handle)
+	cp->handle = handle;
+
+	/* use only enc_size bytes of key for encryption */
+	memcpy(cp->ltk, ltk, sizeof(cp->ltk));
```
The code doesn't seem to match the comment above it. `sizeof(cp->ltk)` will always evaluate to 16, whereas the original code was taking `conn->le.keys->enc_size` into account. So this looks wrong to me.

joerchan replied:
The comment is outdated, I will remove it.
Encryption key size masking should all be handled by `bt_smp_request_ltk`, it is correct to copy the entire array here, because we need to include the already zero-set bits.

joerchan replied:
It's gone now.

joerchan commented on subsys/bluetooth/host/smp.c:
```diff
@@ -308,6 +308,16 @@ static u8_t get_io_capa(void)
+static int smp_send_security_req(struct bt_conn *conn);
+#endif
+
+static bool smp_keys_check(struct bt_conn *conn);
```
Yes, more lines to review though ;)

jhedberg replied:
If you split the reordering into a separate commit it should still be easy to see what the actual changes are.

jhedberg replied:
Ah, I see you already embedded the reordering into the same commit. Let me know what you want to do - to split it out in order to keep the actual changes more visible, or to just merge as is (in which case I'd at least add a comment to the commit message explaining why you're seemingly changing more stuff than necessary)

joerchan replied:
Split out moving of smp_keys_check into it's own commit.

joerchan committed:
Bluetooth: SMP: Move bt_smp_keys_check function up

Move the bt_smp_keys_check function above all usage of the function
in order to avoid prototype declaration when making the function static.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: SMP: Move re-encryption handling into the SMP module

Move initiating security functionality and LTK requesting into the SMP
module so that SMP can track when the connection is in the encryption
process

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
